### PR TITLE
Adding ARM64 support

### PR DESCRIPTION
**Pull request by ktsakalozos**
- Etcd for arm is unsupported/experimental needs an extra flag to start.
- We deployments behind a proxy settings.
- Addon images need to be for both amd64 and arm64.
- Storage host provider needs a build on arm64.
- Tests are passing.

### CODE REVIEW

File: microk8s-resources/actions/disable.dns.sh [hunk 1]
```diff
@@ -15,6 +15,10 @@ sleep 5
 # Apply the dns yaml
 # We do not need to see dns pods running at this point just give some slack
 echo "Removing DNS manifest"
-"$SNAP/kubectl" "--kubeconfig=$SNAP/client.config" "delete" "-f" "${SNAP}/actions/dns.yaml"
+
+ARCH=$(arch)
+cat "${SNAP}/actions/dns.yaml" | \
+"$SNAP/bin/sed" 's@\$ARCH@'"$ARCH"'@g' | \
+"$SNAP/kubectl" "--kubeconfig=$SNAP/client.config" delete -f -
```
Comment (johnsca) [inline microk8s-resources/actions/disable.dns.sh]: Why isn't this usibg `use_manifest`?

File: microk8s-resources/actions/disable.ingress.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+ARCH=$(arch)
+TAG="0.15.0"
+if [ "${ARCH}" = arm64 ]
+then
+  TAG="0.11.0"
+fi
+cat "${SNAP}/actions/ingress.yaml" | \
+"$SNAP/bin/sed" 's@\$ARCH@'"$ARCH"'@g' | \
+"$SNAP/bin/sed" 's@\$TAG@'"$TAG"'@g' | \
```
Comment (johnsca) [inline microk8s-resources/actions/disable.ingress.sh]: Would be nice for `use_manifest` to support this, maybe as an option, just for consistency.  Not sure how to do it in a generic way, though.  Maybe a largish `get_version_tag` helper or something?

File: microk8s-resources/actions/disable.storage.sh [hunk 3]
```diff
@@ -2,9 +2,12 @@
 
 set -e
 
+source $SNAP/actions/common/utils.sh
+
 echo "Disabling default storage"
 cat "${SNAP}/actions/storage.yaml" | \
 "$SNAP/bin/sed" 's@\$SNAP_COMMON@'"$SNAP_COMMON"'@g' | \
+"$SNAP/bin/sed" 's@\$ARCH@'"$ARCH"'@g' | \
```
Comment (johnsca) [inline microk8s-resources/actions/disable.storage.sh]: `use_manifest`?

File: microk8s-resources/actions/enable.dns.sh [hunk 4]
```diff
@@ -8,7 +8,10 @@ source $SNAP/actions/common/utils.sh
 # We do not need to see dns pods running at this point just give some slack
 echo "Enabling DNS"
 echo "Applying manifest"
-"$SNAP/kubectl" "--kubeconfig=$SNAP/client.config" "apply" "-f" "${SNAP}/actions/dns.yaml"
+ARCH=$(arch)
+cat "${SNAP}/actions/dns.yaml" | \
+"$SNAP/bin/sed" 's@\$ARCH@'"$ARCH"'@g' | \
+"$SNAP/kubectl" "--kubeconfig=$SNAP/client.config" apply -f -
```
Comment (johnsca) [inline microk8s-resources/actions/enable.dns.sh]: `use_manifest`?

File: microk8s-resources/actions/enable.ingress.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ARCH=$(arch)
+TAG="0.15.0"
+if [ "${ARCH}" = arm64 ]
+then
+  TAG="0.11.0"
+fi
+cat "${SNAP}/actions/ingress.yaml" | \
+"$SNAP/bin/sed" 's@\$ARCH@'"$ARCH"'@g' | \
+"$SNAP/bin/sed" 's@\$TAG@'"$TAG"'@g' | \
+"$SNAP/kubectl" "--kubeconfig=$SNAP/client.config" apply -f -
```
Comment (johnsca) [inline microk8s-resources/actions/enable.ingress.sh]: `use_manifest`?

File: microk8s-resources/actions/enable.storage.sh [hunk 6]
```diff
@@ -2,9 +2,13 @@
...
 set -e
 
+source $SNAP/actions/common/utils.sh
+
 echo "Enabling default storage class"
 sudo mkdir -p ${SNAP_COMMON}/default-storage
+ARCH=$(arch)
 cat "${SNAP}/actions/storage.yaml" | \
 "$SNAP/bin/sed" 's@\$SNAP_COMMON@'"$SNAP_COMMON"'@g' | \
+"$SNAP/bin/sed" 's@\$ARCH@'"$ARCH"'@g' | \
```
Comment (johnsca) [inline microk8s-resources/actions/enable.storage.sh]: `use_manifest`?

File: microk8s-resources/wrappers/run-etcd-with-args [hunk 7]
```diff
@@ -9,6 +9,15 @@ if [ -S "/var/snap/microk8s/current/etcd.socket:2379" ]; then
...
 fi
 
+export ARCH="${KUBE_ARCH:-`dpkg --print-architecture`}"
+if [ "$ARCH" = "armhf" ]; then
+  ARCH="arm"
+fi
+
+if ! [ "$ARCH" = "amd64" ] && ! [ "$ARCH" = "ppc64el" ]; then
+  export ETCD_UNSUPPORTED_ARCH="$ARCH"
+fi
```
Comment (johnsca) [inline microk8s-resources/wrappers/run-etcd-with-args]: Use `arch` helper here?

File: snap/hooks/configure [hunk 8]
```diff
@@ -36,3 +36,22 @@ if grep -e  "requestheader-client-ca-file=/var/snap/microk8s/.../certs/ca.crt"
...
+# HTTPS_PROXY=http://username:password@proxy:port/ dockerd.
+# where username: and password@ are optional. eg:
+#
+# HTTPS_PROXY=https://squid.internal:3128
+#
+#
+# Remember to restart the docker daemon after editing this file:
+#
+# sudo systemctl restart snap.microk8s.daemon-docker.service
+" > ${SNAP_DATA}/args/dockerd-env
```
Comment (johnsca) [inline snap/hooks/configure]: Shouldn't this copy from `default-args`?

Comment (ktsakalozos) [reply snap/hooks/configure]: Yes! You are right.

File: microk8s-resources/wrappers/run-etcd-with-args (hunk 7 above)
Comment (ktsakalozos) [reply microk8s-resources/wrappers/run-etcd-with-args]: Yes!

**Comment by ktsakalozos** [comment]
@johnsca thank you for the review. I did a refactor on the use_manifest function so that it optionally takes a map of key value pairs to replace on the manifest.

**Comment by ktsakalozos** [comment]
@johnsca I am merging this because we need the work on the proxy. Our new CI is behind a proxy so we need microk8s to work there as well.

**Pull request merged**